gh-90449: Improve accuracy and readability of exceptions tutorial (#31899)
python/cpython: Doc/tutorial/errors.rst

When an exception occurs, it may have associated values, also known as the
exception's *arguments*. The presence and types of the arguments depend on the
exception type.

The *except clause* may specify a variable after the exception name.  The
variable is bound to the exception instance which typically has an ``args``
attribute that stores the arguments. For convenience, builtin exception
types define :meth:`__str__` to print all the arguments without explicitly
accessing ``.args``.  ::

   >>> try:
   ...     raise Exception('spam', 'eggs')
   ... except Exception as inst:
   ...     print(type(inst))    # the exception instance
   ...     print(inst.args)     # arguments stored in .args
   ...     print(inst)          # __str__ allows args to be printed directly,
   ...                          # but may be overridden in exception subclasses
   ...     x, y = inst.args     # unpack args
   ...     print('x =', x)
   ...     print('y =', y)
   ...
   <class 'Exception'>
   ('spam', 'eggs')
   ('spam', 'eggs')
   x = spam
   y = eggs

The exception's :meth:`__str__` output is printed as the last part ('detail')
of the message for unhandled exceptions.

> :exc:`BaseException` is the common base class of all exceptions, including
> :exc:`Exception` which is a subclass of all exceptions apart a few which
> are not typically handled, because they are used to indicate that the
> program should terminate. They include :exc:`SystemExit` which is raised by

Reviewer: "apart" should be "apart from", but the whole sentence is rather complex. Here's a suggested rewrite:

```suggestion
:exc:`Exception` is the common base class of almost all exceptions. It is
itself a subclass of :exc:`BaseException`. A few other exceptions also
derive directly from :exc:`BaseException`. These should not typically be
handled, because they indicate that the program should terminate. They
include :exc:`SystemExit` which is raised by
```

This avoids overly long sentences and gives prominence to `Exception`, which I think is more important to cover in the tutorial.

Author: I don't object to rewording this sentence, but I don't think the suggested change works.  This doesn't really make sense at this point:

"A few other exceptions also derive directly from :exc:`BaseException`, because they indicate that the program should terminate and should not typically be handled."

because it's not clear why the fact that "they indicate .. and should not be handled" explain why they are not BaseException subclasses.

Reviewer: Good point, reworded my suggestion above

Author: That's better, but we are no longer saying that BaseException is the base class of all exceptions.

Reviewer: True, but do we need to? This is the tutorial, not the language reference.

Author: I don't think it's clear why BaseException exists without that point.

Author: I've reworded this paragraph, hopefully the new version is better.

When an exception occurs, it may have associated values, also known as the
exception's *arguments*. The presence and types of the arguments depend on the
exception type.

The *except clause* may specify a variable after the exception name.  The
variable is bound to the exception instance which typically has an ``args``
attribute that stores the arguments. For convenience, builtin exception
types define :meth:`__str__` to print all the arguments without explicitly
accessing ``.args``.  ::

   >>> try:
   ...     raise Exception('spam', 'eggs')
   ... except Exception as inst:
   ...     print(type(inst))    # the exception instance
   ...     print(inst.args)     # arguments stored in .args
   ...     print(inst)          # __str__ allows args to be printed directly,
   ...                          # but may be overridden in exception subclasses
   ...     x, y = inst.args     # unpack args
   ...     print('x =', x)
   ...     print('y =', y)
   ...
   <class 'Exception'>
   ('spam', 'eggs')
   ('spam', 'eggs')
   x = spam
   y = eggs

The exception's :meth:`__str__` output is printed as the last part ('detail')
of the message for unhandled exceptions.

> :exc:`BaseException` is a common base class of all exceptions, including a few

Reviewer 2: ```suggestion
:exc:`BaseException` is the common base class of all exceptions, including a few
```

which are not typically handled, because they are used to indicate that the

> program should terminate. They include :exc:`SystemExit` which is raise by

Reviewer 2: ```suggestion
program should terminate. They include :exc:`SystemExit` which is raised by
```

> ``sys.exit()`` and :exc:`KeyboardInterrupt` which is raised when a user

Reviewer 2: Is `sys.exit` not a link on purpose?

> ``sys.exit()`` and :exc:`KeyboardInterrupt` which is raised when a user

Author: ```suggestion
:meth:`sys.exit` and :exc:`KeyboardInterrupt` which is raised when a user
```

> attempts wishes to interrupt the program.

Reviewer 2: This sentence has has two verbs

> attempts wishes to interrupt the program.

Author: ```suggestion
wishes to interrupt the program.
```

:exc:`Exception` is a subclass of :exc:`BaseException`, which is a common
superclass of all exceptions, apart from those few special ones that we

> should not typically handle. As such, :exc:`Exception` can be used as a

Reviewer 2: This repeats lines 180-184, should probably be adapted before merge (no rush!)

Author: 180-184 are talking about BaseException, and this is about Exception.

Reviewer 2: Ah, I read that as an apposition `BaseException, which is a common etc`, so it is ambiguous.

Author: Ah after re-reading I see the repetition, will fix!

Note that if the *except clauses* were reversed (with ``except B`` first), it
would have printed B, B, B --- the first matching *except clause* is triggered.

> When an exception occurs, it may have associated values, also known as the

Author: This paragraph needed to move up because the next example uses a named except clause. I also reworded it a bit.

exception's *arguments*. The presence and types of the arguments depend on the
exception type.

The *except clause* may specify a variable after the exception name.  The
variable is bound to the exception instance which typically has an ``args``
attribute that stores the arguments. For convenience, builtin exception
types define :meth:`__str__` to print all the arguments without explicitly
accessing ``.args``.  ::

   >>> try:
   ...     raise Exception('spam', 'eggs')
   ... except Exception as inst:
   ...     print(type(inst))    # the exception instance
   ...     print(inst.args)     # arguments stored in .args
   ...     print(inst)          # __str__ allows args to be printed directly,
   ...                          # but may be overridden in exception subclasses
   ...     x, y = inst.args     # unpack args
   ...     print('x =', x)
   ...     print('y =', y)
   ...
   <class 'Exception'>
   ('spam', 'eggs')
   ('spam', 'eggs')
   x = spam
   y = eggs

The exception's :meth:`__str__` output is printed as the last part ('detail')
of the message for unhandled exceptions.

All exceptions inherit from :exc:`BaseException`, and so it can be used to serve

> as a wildcard. Use this with extreme caution, since it is easy to mask a real

Reviewer 2: Not a fan of this section!  `except Exception` is the catch-all recommended by style guides and linting tools; you have to know what you’re doing when you write `except BaseException`.

I saw this was discussed in the ticket but didn’t read all messages.  Do you think it’s in scope for this ticket, or should be a discussion somewhere with a wider core dev audience?

Author: We can discuss it here I think. I agree it's a dodgy thing to have in a tutorial.  And we probably don't want to talk about the no-type except: clause here either.

       s = f.readline()
       i = int(s.strip())
   except OSError as err:

>        print(f"OS error: {err}")

Reviewer 2: ```suggestion
       print("OS error:", err)
```

I often see that people love f-strings and forget that `print` is amazing!  Accepts multiple params, converts them to strings, adds a space or something else between them, all that.

   except ValueError:
       print("Could not convert data to an integer.")
   except Exception as err:

>        print(f"Unexpected {err=}, {type(err)=}")

Reviewer 2: Personnally I find this formatting a little advanced for a tuto, but it was here before so it is what it is.

that wasn't raised by the code being protected by the :keyword:`!try` ...
:keyword:`!except` statement.

Exception handlers do not handle only exceptions that occur immediately in the
*try clause*, but also those that occur inside functions that are called (even

> indirectly) from the *try clause*. For example::

Reviewer 2: I don’t think the previous use of `in` was incorrect or confusing, and don’t really like `from` here.

> indirectly) from the *try clause*. For example::

Author: ```suggestion
indirectly) in the *try clause*. For example::
```

The sole argument to :keyword:`raise` indicates the exception to be raised.
This must be either an exception instance or an exception class (a class that

> derives from :class:`BaseException`).  If an exception class is passed, it will

Reviewer 2: Because we are in the tutorial, I wonder if we should say: `a class that derives from Exception or BaseException`.

Yes, BaseException implies Exception, but it may not be self-evident at this point, so this wording would be technically correct (exception = class deriving from BaseException) and friendly (you should remember Exception!)

Author: Maybe 'a class that derive from BaseException, such as Exception or one if its subclasses' ?

naming of the standard exceptions.

Many standard modules define their own exceptions to report errors that may

> occur in functions they define.

Author: This is mentioned above as well, so I removed the repetition.
